chia/wallet/util/wallet_sync_utils.py

```diff
@@ -389,3 +390,29 @@ async def fetch_header_blocks_in_range(
         assert res_h_blocks is not None
         blocks.extend([bl for bl in res_h_blocks.header_blocks if bl.height >= start])
     return blocks
+
+
+async def fetch_puzzle_solution_with_coin_info(height: uint32, coin: Coin, peer: WSChiaConnection) -> CoinSpend:
+    if Capability.PUZZLE_SOLUTION_WITH_COIN_INFO in peer.peer_capabilities:
+        solution_response = await peer.request_puzzle_solution_with_coin_info(
+            wallet_protocol.RequestPuzzleSolutionWithCoinInfo(
+                bytes32(coin.parent_coin_info),
+                bytes32(coin.puzzle_hash),
+                uint64(coin.amount),
+                height,
+            )
+        )
+    else:
+        solution_response = await peer.request_puzzle_solution(
+            wallet_protocol.RequestPuzzleSolution(coin.name(), height)
+        )
+    if solution_response is None or not isinstance(solution_response, wallet_protocol.RespondPuzzleSolution):
+        raise ValueError(f"Was not able to obtain solution {solution_response}")
+    assert solution_response.response.puzzle.get_tree_hash() == coin.puzzle_hash
+    assert solution_response.response.coin_name == coin.name()
```

Comment:
I would prefer we move away from using `assert` as a data sanitization mechanism. Normally `assert` describes a condition that's always true, short of bugs.